river/Cursor.zig

```diff
@@ -185,15 +186,43 @@ const Mode = union(enum) {
 
     /// Pass an event on to the surface under the cursor, if any.
     fn passthrough(self: *Self, time: u32) void {
+        // Get the output where the cursor is
+        const root = &self.seat.input_manager.server.root;
+        const config = self.seat.input_manager.server.config;
+
+        var ox = self.wlr_cursor.x;
+        var oy = self.wlr_cursor.y;
+        // TODO should we rellly return here? Maybe catch unreachable
+        const output = Output.outputAt(root, &ox, &oy) orelse return;
+
         var sx: f64 = undefined;
         var sy: f64 = undefined;
-        if (self.surfaceAt(self.wlr_cursor.x, self.wlr_cursor.y, &sx, &sy)) |wlr_surface| {
+        if (output.surfaceAt(ox, oy, &sx, &sy)) |wlr_surface| {
             // If input is allowed on the surface, send pointer enter and motion
             // events. Note that wlroots won't actually send an enter event if
             // the surface has already been entered.
             if (self.seat.input_manager.inputAllowed(wlr_surface)) {
+                const wlr_seat = self.seat.wlr_seat;
+                const focus_change = wlr_seat.pointer_state.focused_surface != wlr_surface;
+
                 c.wlr_seat_pointer_notify_enter(self.seat.wlr_seat, wlr_surface, sx, sy);
                 c.wlr_seat_pointer_notify_motion(self.seat.wlr_seat, time, sx, sy);
+                if (View.fromWlrSurface(wlr_surface)) |view| {
+                    // Change focus according to config
+                    switch (config.focus_follows_cursor) {
+                        .disabled => {},
+                        .normal => {
+                            if (focus_change) self.seat.focus(view);
+                            root.startTransaction();
```

Comment:
Let's only call `startTransaction()` if focus is actually set.

Reply:
Fixed.